host/src/gap.rs

```diff
@@ -0,0 +1,122 @@
+//! ## Generic Access Profile
+//!
+//! This profile defines the generic procedures related to discovery of
+//! Bluetooth devices (idle mode procedures) and link management aspects
+//! of connecting to Bluetooth devices (connecting mode procedures).
+//! It also defines procedures related to use of different security levels.
+//! In addition, this profile includes common format requirements for
+//! parameters accessible on the user interface level.
+
+use crate::prelude::*;
+use embassy_sync::blocking_mutex::raw::RawMutex;
+
+pub mod appearance {
+    //! The representation of the external appearance of the device.
+    //!
+    //! This is a list of some of the most common appearance values and demonstrates the pattern to use to define new appearance values.
+    //!
+    //! https://www.bluetooth.com/wp-content/uploads/Files/Specification/Assigned_Numbers.html#bookmark49
+
+    // TODO: Perhaps this should be it's own crate in future?
+
+    /// Construct a new appearance value for the GAP Service.
+    ///
+    /// Follow the pattern of the examples below to create new appearance values.
+    /// Use UUIDs from the [Bluetooth Assigned Numbers list](https://www.bluetooth.com/wp-content/uploads/Files/Specification/Assigned_Numbers.html#bookmark49).
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// use trouble_host::prelude::*;
+    ///
+    /// const GAMEPAD: &[u8; 2] = &appearance::new(0x00F, 0x040);
+    /// ```
+    pub const fn new(category: u8, subcategory: u8) -> [u8; 2] {
+        (((category as u16) << 6) | (subcategory as u16)).to_le_bytes()
+    }
+
+    /// Generic Unknown device appearance.
+    pub const GENERIC_UNKNOWN: [u8; 2] = new(0x000, 0x000);
+    /// Generic Phone device appearance.
+    pub const GENERIC_PHONE: [u8; 2] = new(0x001, 0x000);
+    /// Generic Computer device appearance.
+    pub const GENERIC_COMPUTER: [u8; 2] = new(0x002, 0x000);
+    /// Smart Watch device appearance.
+    pub const SMART_WATCH: [u8; 2] = new(0x003, 0x020);
+    /// Generic Power device appearance.
+    pub const GENERIC_POWER: [u8; 2] = new(0x01E, 0x000);
+    /// Generic Sensor device appearance.
+    pub const GENERIC_SENSOR: [u8; 2] = new(0x015, 0x000);
+    /// Gamepad device appearance.
+    pub const GAMEPAD: [u8; 2] = new(0x00F, 0x040);
+}
+
+/// Configuration for the GAP Service.
+pub enum GapConfig {
+    /// Peripheral device configuration.
+    Peripheral(PeripheralConfig),
+    /// Central device configuration.
+    Central(CentralConfig),
+}
+
+/// Configuration for a peripheral device GAP Service.
+pub struct PeripheralConfig {
+    /// The name of the peripheral device.
+    pub name: &'static str,
```

Comment:
Would it make sense to 

* Add lifetimes to allow getting values from a stack buffer
* Use &[u8] instead of &str, in cases where you don't want to decode utf-8 in order to pass these values.

Reply:
>     * Add lifetimes to allow getting values from a stack buffer

Yes, that is important since some devices define their name at runtime. Also, there needs to be a way to change the device name after the server has been started.

>     * Use &[u8] instead of &str, in cases where you don't want to decode utf-8 in order to pass these values.

The spec requires UTF-8 so I think `&str` is appropriate here.

Reply:
ok I'll see what I can do about the lifetimes, I had some challenges with them and the attribute table.

Reply:
I don't think its possible to change the device name after the server has been started in the current implementation. I think to do that we'd need to have it as a read and write characteristic, and then perhaps expose a specific method for changing the id.  That might need to be a future PR.

Reply:
I might need a hand with making those lifetimes non-static, I couldn't get the borrow checker to play nice with 'value

Reply:
I'm fine merging this as is now. We can address these points later:

* Moving BLE spec numbers to the bt-hci crate (https://github.com/embassy-rs/trouble/issues/145)
* Supporting non-static lifetimes for gap config (https://github.com/embassy-rs/trouble/issues/146)